consensus/system_contract/consensus.go

```diff
@@ -86,13 +86,18 @@ func (s *SystemContract) VerifyHeader(chain consensus.ChainHeaderReader, header
 // concurrently. The method returns a quit channel to abort the operations and
 // a results channel to retrieve the async verifications (the order is that of
 // the input slice).
-func (s *SystemContract) VerifyHeaders(chain consensus.ChainHeaderReader, headers []*types.Header, seals []bool) (chan<- struct{}, <-chan error) {
+func (s *SystemContract) VerifyHeaders(chain consensus.ChainHeaderReader, headers []*types.Header, seals []bool, parent *types.Header) (chan<- struct{}, <-chan error) {
 	abort := make(chan struct{})
 	results := make(chan error, len(headers))
 
 	go func() {
 		for i, header := range headers {
-			err := s.verifyHeader(chain, header, headers[:i])
+			parents := headers[:i]
+			if len(parents) == 0 && parent != nil {
+				parents = []*types.Header{parent}
+			}
+
+			err := s.verifyHeader(chain, header, parents)
```

Comment:
A clarification comment: it seems all the paths of `clique` and `system_contract` in verify header all use the last header of the parents array.

Reply:
Normally when `BlockChain` calls `VerifyHeaders`, the previous batch (including the parent of the first header) have already been processed and written into the chain.

But during the transition, inside the upgradable consensus engine, we internally call `VerifyHeaders` twice. The problem is that we did not pass the parent to the second invocation (`system_contract`), and the parent was not written into the chain either.

Reply:
Got it. And the clarification is that it seems in both cases (`clique` and `system_contract`) of calling (`verifyHeader`), only passing `parent` instead of an array should be sufficient, seeing no place that needs an array. Not a comment for change interface, just a clarification, and for ethhash engine, `verifyHeader` uses `parent` header instead of an array, too.